topostats/tracing/skeletonize.py

```diff
@@ -2,78 +2,641 @@
+            self.p6 * self.p8 * self.p2,
+            self.p8 * self.p2 * self.p4,
+        )
+
+    @staticmethod
+    def getLocalPixelsBinary(binary_map, x, y) -> np.ndarray:
+        """Get the values of the pixels in the local 8-connectivit area around
+        the coordinate described by x and y.
+
+        Parameters
+        ----------
+        binary_map: np.ndarray
+            The binary array containing the grains.
+        x: int
+            An x coordinate within the binary map.
+        y: int
+            A y coordinate within the binary map.
+
+        Returns
+        -------
+        np.ndarray
+            A flattened 8-long array describing the values in the binary map
+            around the x,y point
+        """
+        local_pixels = binary_map[x - 1 : x + 2, y - 1 : y + 2].flatten()
+        # [[p7, p8, p9],    [[0,1,2],
+        #  [p6, na, p2], ->  [3,4,5], -> [0,1,2,3,5,6,7,8]
+        #  [p5, p4, p3]]     [6,7,8]]
+        # delete coordinate pixel to only get local area
+        return np.delete(local_pixels, 4)
+
+
+class pruneSkeleton:
+    """Class containing skeletonization pruning code from factory methods to functions
+    depaendant on the method. Pruning is the act of removing spurious branches commonly
+    found when implimenting skeletonization algorithms."""
+
+    def __init__(self, image: np.ndarray, skeleton: np.ndarray) -> None:
+        """Initialise the class.
+
+        Parameters
+        ----------
+        image: np.ndarray
+            The original image the skeleton derives from (not the binary mask)
+        skeleton: np.ndarray
+            The single-pixel-thick skeleton pertaining to features of the image.
+        """
+        self.image = image
+        self.skeleton = skeleton
+
+    def prune_skeleton(self, method: str = "joe") -> np.ndarray:
+        """Factory method for pruning skeletons.
+
+        Parameters
+        ----------
+        method : str
+            Method to use, default is 'joe'.
+
+        Returns
+        -------
+        np.ndarray
+            An array of the skeleton with spurious branching artefacts removed.
+
+        Notes
+        -----
+
+        This is a thin wrapper to the methods provided within the pruning classes below.
+        """
+        return self._prune_method(method)
+
+    def _prune_method(self, method: str = "joe") -> Callable:
+        """Creator component which determines which skeletonize method to use.
+
+        Parameters
+        ----------
+        method: str
+            Method to use for skeletonizing, methods are 'joe' other options are 'conv'.
+
+        Returns
+        -------
+        Callable
+            Returns the function appropriate for the required skeletonizing method.
+        """
+        if method == "joe":
+            return self._prune_joe(self.image, self.skeleton)
+        raise ValueError(method)
+
+    @staticmethod
+    def _prune_joe(image: np.ndarray, skeleton: np.ndarray) -> np.ndarray:
```

Comment:
I think a more descriptive name here would be appropriate. `joe` doesn't mean much to someone coming to this with fresh eyes.